File: src/python/xerxes/runtime/rate_limit_tracker.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Track per-(provider, model) rate-limit state from response headers.

Parses ``x-ratelimit-*`` and ``retry-after`` headers, retains the latest
snapshot per (provider, model) pair, and exposes :meth:`RateLimitTracker.should_throttle`
plus :meth:`RateLimitTracker.delay_ms` so callers can pre-emptively pause
before they get a 429. Defaults align with the OpenAI / Anthropic header
conventions and the constant tuples on :class:`RateLimitTracker` make it
straightforward to add new dialects.
"""

from __future__ import annotations

import threading
import time
import typing as tp
from dataclasses import dataclass, field


def _parse_int(value: tp.Any) -> int | None:
    """Parse ``value`` as an ``int`` or return ``None`` for non-numeric input."""
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _parse_float(value: tp.Any) -> float | None:
    """Parse ``value`` as a ``float`` or return ``None`` for non-numeric input."""
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


@dataclass
class RateLimitState:
    """Latest snapshot of the rate-limit budget for a given provider+model.

    Attributes:
        limit_requests: Total requests-per-minute budget reported by the provider.
        remaining_requests: Requests still available in the current window.
        limit_tokens: Total tokens-per-minute budget.
        remaining_tokens: Tokens still available in the current window.
        reset_at: Epoch seconds when the request bucket resets, if known.
        retry_after: Explicit retry hint in seconds from the most recent 429.
        last_updated: Wall-clock time when this snapshot was last refreshed.
    """

    limit_requests: int | None = None
    remaining_requests: int | None = None
    limit_tokens: int | None = None
    remaining_tokens: int | None = None
    reset_at: float | None = None
    retry_after: float | None = None
    last_updated: float = field(default_factory=time.time)


class RateLimitTracker:
    """Track and reason about LLM-API rate limits per (provider, model)."""

    # Standard header names — lowercase, with fallbacks for common dialects.
    HEADER_LIMIT_REQS = ("x-ratelimit-limit-requests", "x-ratelimit-limit")
    HEADER_REMAINING_REQS = ("x-ratelimit-remaining-requests", "x-ratelimit-remaining")
    HEADER_LIMIT_TOKENS = ("x-ratelimit-limit-tokens",)
    HEADER_REMAINING_TOKENS = ("x-ratelimit-remaining-tokens",)
    HEADER_RESET = ("x-ratelimit-reset-requests", "x-ratelimit-reset")
    HEADER_RETRY_AFTER = ("retry-after",)

    def __init__(self, *, throttle_ratio: float = 0.05) -> None:
        """Construct a new tracker with a configurable throttle threshold.

        Args:
            throttle_ratio: Fraction of the limit below which
                :meth:`should_throttle` flips to ``True``. Must be in
                ``(0.0, 1.0)``; defaults to a conservative ``0.05`` (5%).

        Raises:
            ValueError: ``throttle_ratio`` is outside ``(0, 1)``.
        """
        if not 0.0 < throttle_ratio < 1.0:
            raise ValueError("throttle_ratio must be in (0.0, 1.0)")
        self._throttle_ratio = throttle_ratio
        self._states: dict[tuple[str, str], RateLimitState] = {}
        self._lock = threading.Lock()

    @staticmethod
    def _pluck(headers: dict[str, str], names: tuple[str, ...]) -> str | None:
        """Case-insensitive lookup that tries each name in order."""
        lower = {k.lower(): v for k, v in headers.items()}
        for n in names:
            v = lower.get(n)
            if v is not None:
                return v
        return None

    def update(self, provider: str, model: str, headers: dict[str, str]) -> RateLimitState:
        """Merge a response's rate-limit headers into the tracker.

        Args:
            provider: Provider identifier, e.g. ``"openai"`` or ``"anthropic"``.
            model: Model identifier as reported by the API.
            headers: Raw response headers; lookups are case-insensitive.

        Returns:
            The updated :class:`RateLimitState` for the ``(provider, model)`` key.
            Missing headers are left untouched (a partial update never wipes
            an earlier observation).
        """
        now = time.time()
        with self._lock:
            state = self._states.setdefault((provider, model), RateLimitState())
            v = self._pluck(headers, self.HEADER_LIMIT_REQS)
            if (parsed := _parse_int(v)) is not None:
                state.limit_requests = parsed
            v = self._pluck(headers, self.HEADER_REMAINING_REQS)
            if (parsed := _parse_int(v)) is not None:
                state.remaining_requests = parsed
            v = self._pluck(headers, self.HEADER_LIMIT_TOKENS)
            if (parsed := _parse_int(v)) is not None:
                state.limit_tokens = parsed
            v = self._pluck(headers, self.HEADER_REMAINING_TOKENS)
            if (parsed := _parse_int(v)) is not None:
                state.remaining_tokens = parsed
            v = self._pluck(headers, self.HEADER_RESET)
            if (parsed := _parse_float(v)) is not None:
                state.reset_at = now + parsed
            v = self._pluck(headers, self.HEADER_RETRY_AFTER)
            if (parsed := _parse_float(v)) is not None:
                state.retry_after = parsed
            state.last_updated = now
            return state

    def state(self, provider: str, model: str) -> RateLimitState | None:
        """Return the latest snapshot for the key, or ``None`` if never seen."""
        with self._lock:
            return self._states.get((provider, model))

    def should_throttle(self, provider: str, model: str) -> bool:
        """Return ``True`` when the request or token budget is dangerously low.

        Also returns ``True`` while a recent ``retry-after`` window is still
        active.
        """
        st = self.state(provider, model)
        if st is None:
            return False
        if st.retry_after and st.last_updated + st.retry_after > time.time():
            return True
        # Check both ratios — whichever signals first wins.
        for limit, remaining in ((st.limit_requests, st.remaining_requests), (st.limit_tokens, st.remaining_tokens)):
            if limit and remaining is not None and limit > 0:
                ratio = remaining / limit
                if ratio < self._throttle_ratio:
                    return True
        return False

    def delay_ms(self, provider: str, model: str) -> int:
        """Return a suggested pre-call delay in milliseconds.

        Returns ``0`` when no throttling is needed. Honours a still-active
        ``retry_after`` first; otherwise sleeps up to the bucket reset (capped
        at 60 seconds) when :meth:`should_throttle` is true, falling back to a
        small conservative default when no reset hint is available.
        """
        st = self.state(provider, model)
        if st is None:
            return 0
        if st.retry_after:
            remaining_window = st.last_updated + st.retry_after - time.time()
            if remaining_window > 0:
                return int(remaining_window * 1000)
        if not self.should_throttle(provider, model):
            return 0
        # If budget is very low, sleep until reset (capped at 60 s).
        if st.reset_at:
            until_reset = max(0.0, st.reset_at - time.time())
            return int(min(60.0, until_reset) * 1000)
        return 250  # conservative default

    def clear(self) -> None:
        """Drop every tracked ``(provider, model)`` snapshot."""
        with self._lock:
            self._states.clear()


__all__ = ["RateLimitState", "RateLimitTracker"]

File: src/typescript/src/runtime/rateLimitTracker.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export type RateLimitHeaders = Headers | Readonly<Record<string, string | number | undefined>>

/** Latest rate-limit observation for one provider/model pair. */
export class RateLimitState {
  lastUpdated: number
  limitRequests: number | undefined
  limitTokens: number | undefined
  remainingRequests: number | undefined
  remainingTokens: number | undefined
  resetAt: number | undefined
  retryAfter: number | undefined
  retryAfterUntil: number | undefined

  constructor(now = Date.now() / 1_000) {
    this.lastUpdated = now
  }
}

export interface RateLimitTrackerOptions {
  /** Fraction of a known budget below which pre-emptive throttling starts. */
  readonly throttleRatio?: number
  /** Epoch-seconds clock; injectable for deterministic callers and tests. */
  readonly now?: () => number
}

/**
 * Tracks provider rate-limit headers per provider/model pair and offers
 * deterministic throttle and delay decisions before callers hit a 429.
 */
export class RateLimitTracker {
  static readonly HEADER_LIMIT_REQUESTS = ['x-ratelimit-limit-requests', 'x-ratelimit-limit'] as const
  static readonly HEADER_REMAINING_REQUESTS = ['x-ratelimit-remaining-requests', 'x-ratelimit-remaining'] as const
  static readonly HEADER_LIMIT_TOKENS = ['x-ratelimit-limit-tokens'] as const
  static readonly HEADER_REMAINING_TOKENS = ['x-ratelimit-remaining-tokens'] as const
  static readonly HEADER_RESET = ['x-ratelimit-reset-requests', 'x-ratelimit-reset'] as const
  static readonly HEADER_RETRY_AFTER = ['retry-after'] as const
  /** Python-compatible abbreviation retained at the public boundary. */
  static readonly HEADER_LIMIT_REQS = RateLimitTracker.HEADER_LIMIT_REQUESTS
  /** Python-compatible abbreviation retained at the public boundary. */
  static readonly HEADER_REMAINING_REQS = RateLimitTracker.HEADER_REMAINING_REQUESTS

  private readonly clock: () => number
  private readonly states = new Map<string, RateLimitState>()
  private readonly throttleRatio: number

  constructor(options: RateLimitTrackerOptions | number = {}) {
    const normalized = typeof options === 'number' ? { throttleRatio: options } : options
    const throttleRatio = normalized.throttleRatio ?? 0.05
    if (!Number.isFinite(throttleRatio) || throttleRatio <= 0 || throttleRatio >= 1) {
      throw new RangeError('throttleRatio must be in (0.0, 1.0)')
    }
    this.throttleRatio = throttleRatio
    this.clock = normalized.now ?? (() => Date.now() / 1_000)
  }

  /** Merge response headers into a provider/model snapshot without clearing absent values. */
  update(provider: string, model: string, headers: RateLimitHeaders, now = this.clock()): RateLimitState {
    const state = this.states.get(stateKey(provider, model)) ?? new RateLimitState(now)
    const normalized = normalizeHeaders(headers)
    assignInteger(state, 'limitRequests', pluck(normalized, RateLimitTracker.HEADER_LIMIT_REQUESTS))
    assignInteger(state, 'remainingRequests', pluck(normalized, RateLimitTracker.HEADER_REMAINING_REQUESTS))
    assignInteger(state, 'limitTokens', pluck(normalized, RateLimitTracker.HEADER_LIMIT_TOKENS))
    assignInteger(state, 'remainingTokens', pluck(normalized, RateLimitTracker.HEADER_REMAINING_TOKENS))

    const reset = parseFloatValue(pluck(normalized, RateLimitTracker.HEADER_RESET))
    if (reset !== undefined) state.resetAt = now + reset
    const retryAfter = parseFloatValue(pluck(normalized, RateLimitTracker.HEADER_RETRY_AFTER))
    if (retryAfter !== undefined) {
      state.retryAfter = retryAfter
      state.retryAfterUntil = now + retryAfter
    }
    state.lastUpdated = now
    this.states.set(stateKey(provider, model), state)
    return state
  }

  /** Return a provider/model snapshot, or undefined when no response has supplied headers. */
  state(provider: string, model: string): RateLimitState | undefined {
    return this.states.get(stateKey(provider, model))
  }

  /** Return true while Retry-After is active or a known request/token budget is nearly exhausted. */
  shouldThrottle(provider: string, model: string, now = this.clock()): boolean {
    const state = this.state(provider, model)
    if (!state) return false
    if (state.retryAfterUntil !== undefined && state.retryAfterUntil > now) return true
    return (
      this.budgetIsLow(state.limitRequests, state.remainingRequests)
      || this.budgetIsLow(state.limitTokens, state.remainingTokens)
    )
  }

  /**
   * Return a pre-call delay in milliseconds. Retry-After wins; otherwise a
   * throttled bucket waits until reset (capped at 60 seconds) or 250ms.
   */
  delayMs(provider: string, model: string, now = this.clock()): number {
    const state = this.state(provider, model)
    if (!state) return 0
    if (state.retryAfterUntil !== undefined) {
      const remaining = state.retryAfterUntil - now
      if (remaining > 0) return Math.trunc(remaining * 1_000)
    }
    if (!this.shouldThrottle(provider, model, now)) return 0
    if (state.resetAt !== undefined) return Math.trunc(Math.min(60, Math.max(0, state.resetAt - now)) * 1_000)
    return 250
  }

  /** Drop every provider/model snapshot. */
  clear(): void {
    this.states.clear()
  }

  private budgetIsLow(limit: number | undefined, remaining: number | undefined): boolean {
    return limit !== undefined && remaining !== undefined && limit > 0 && remaining / limit < this.throttleRatio
  }
}

function assignInteger(
  state: RateLimitState,
  field: 'limitRequests' | 'remainingRequests' | 'limitTokens' | 'remainingTokens',
  raw: string | undefined,
): void {
  const parsed = parseInteger(raw)
  if (parsed !== undefined) state[field] = parsed
}

function normalizeHeaders(headers: RateLimitHeaders): Map<string, string> {
  const normalized = new Map<string, string>()
  if (headers instanceof Headers) {
    for (const [name, value] of headers.entries()) normalized.set(name.toLowerCase(), value)
    return normalized
  }
  for (const [name, value] of Object.entries(headers)) {
    if (value !== undefined) normalized.set(name.toLowerCase(), String(value))
  }
  return normalized
}

function pluck(headers: ReadonlyMap<string, string>, names: readonly string[]): string | undefined {
  for (const name of names) {
    const value = headers.get(name)
    if (value !== undefined) return value
  }
  return undefined
}

function parseInteger(value: string | undefined): number | undefined {
  if (value === undefined || !/^[+-]?\d+$/.test(value.trim())) return undefined
  const parsed = Number(value)
  return Number.isSafeInteger(parsed) ? parsed : undefined
}

function parseFloatValue(value: string | undefined): number | undefined {
  if (value === undefined || value.trim() === '') return undefined
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : undefined
}

function stateKey(provider: string, model: string): string {
  return `${provider}\u0000${model}`
}
